src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+developed is a good thing
+
+.. todo:: What about functional tests
+
+Continuous Integration
+----------------------
+GitLab manages builds via a YAML file held inside your repository. Similar to a
+Jenkinsfile in philosophy. This is the file in the skeleton HelloWorld project. 
+
+.. literalinclude:: example-ci.yml 
+
+We have four stages of the CI 
+
+* build
+* lint
+* test
+* pages
+
+These stages are automatically run by the GitLab runners when you push to the repository. The pipeline halts and you are informed if any of the steps fail. There are some subtleties in the way the test results and test coverage are reported and we deal with them below as we go through the steps in more detail.
+
+Building The Project
+^^^^^^^^^^^^^^^^^^^^
+.. todo:: Need to add some CMake Coding Conventions
+
+The Image
+"""""""""
+We recommend building using the cpp_base image that we have developed and
+stored in the image repository. This, or an image derived from it contains all
+the tools required to operate this CI pipeline and you should avoid the pain of
+installing and configuring dependencies.
+
+
+CMake Coding Conventions
+""""""""""""""""""""""""
+We recommend using CMake as a build tool. It is widely used, includes the
+ability to generate buildfiles for different development environments. 
+This allows developers the freedom to use whatever IDE they
+would like (e.g. Eclipse and Xcode) from the same CMake files.
+
+Unfortunately there is even less consensus in the community as to the most
+effective way to write CMake files, thought there are a lot of opinions. We
+have found a good distillation of some effective ideas can be found `here
+<https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1>`_. And GitLab
+has an introduction to `Modern CMake
+<https://cliutils.gitlab.io/modern-cmake/>`_.
+
+In general however we recommend Modern CMake (minimum version of 3.0). A lot of legacy
+projects will have used 2.8 - but we cannot recommend that new projects use
+this. It is no longer maintained and does not contain many features of modern
+(v3+ CMake). As far as we are aware CMake v3.0 is backward compatible
+with 2.8, so if you are onboarding code that uses 2.8 we strongly suggest you upgrade. 
+
+.. todo:: Maybe add some more advice? use targets not global settings etc ...
+
+Including the version number
+""""""""""""""""""""""""""""
+There is a template configuration file that demonstrates how to pull in the
+version number into the code base. We do this in a few  steps. First the Semantic
+Version number is held in plain text in version.txt. This is done so that it is
+easy to "bump" the version number on code changes. The second stage is the the
+CMakeLists.txt file reads this file and creates an internal version number
+using its contents. Finally the configuration file is used to generate a header
+file containing the version number that can be included by the code.
+
+Including External Projects
+"""""""""""""""""""""""""""
```

Comment:
this section needs a lot more work. It implies that this is the way we want to do dependencies and as per the JIRa discussion it really isn't - gtest is the exception to the rule.

Reply:
I've kept this the way it was - I think we need a specific feature in PI4 to deal with dependencies